File: SDKs/java/src/main/java/com/intelliversex/sdk/discord/IVXDiscordSocial.java
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

package com.intelliversex.sdk.discord;

import com.intelliversex.sdk.discord.IVXDiscordModels.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;
import java.util.logging.Logger;

/**
 * Thread-safe singleton for the IntelliVerseX Discord Social SDK integration.
 * <p>
 * Wraps Discord Rich Presence, a unified friends list, lobby/text-chat,
 * voice channels, and game invites behind a single ergonomic API.
 * All async operations return {@link CompletableFuture}. Event callbacks
 * are delivered via {@link Consumer} listeners.
 * <p>
 * <b>Usage:</b>
 * <pre>{@code
 * IVXDiscordConfig config = new IVXDiscordConfig("app-id", "client-id", null, true);
 * IVXDiscordSocial.getInstance().initialize(config);
 * IVXDiscordSocial.getInstance().setActivity("Playing QuizVerse", "Round 3");
 * }</pre>
 */
public class IVXDiscordSocial {

    private static final Logger LOG = Logger.getLogger(IVXDiscordSocial.class.getName());

    private static volatile IVXDiscordSocial instance;

    private volatile boolean initialized;
    private IVXDiscordConfig config;
    private IVXDiscordLobbyInfo currentLobby;
    private final List<ChatEntry> chatHistory = new CopyOnWriteArrayList<>();
    private final ConcurrentHashMap<String, List<Consumer<Object>>> listeners = new ConcurrentHashMap<>();

    private IVXDiscordSocial() {}

    /**
     * Returns the singleton instance, creating it on first access (double-checked locking).
     *
     * @return the shared {@link IVXDiscordSocial} instance
     */
    public static IVXDiscordSocial getInstance() {
        if (instance == null) {
            synchronized (IVXDiscordSocial.class) {
                if (instance == null) {
                    instance = new IVXDiscordSocial();
                }
            }
        }
        return instance;
    }

    // ──────────────────────────────────────────────
    //  Lifecycle
    // ──────────────────────────────────────────────

    /**
     * Initialize the Discord Social SDK with the given configuration.
     * Must be called before any other method.
     *
     * @param config Discord application and OAuth configuration
     * @throws IllegalArgumentException if applicationId or clientId is null/blank
     */
    public synchronized void initialize(IVXDiscordConfig config) {
        if (config == null) {
            throw new IllegalArgumentException("config must not be null");
        }
        if (config.getApplicationId() == null || config.getApplicationId().trim().isEmpty()) {
            throw new IllegalArgumentException("applicationId must not be null or empty");
        }
        if (config.getClientId() == null || config.getClientId().trim().isEmpty()) {
            throw new IllegalArgumentException("clientId must not be null or empty");
        }
        this.config = config;
        this.initialized = true;
        log("Discord Social SDK initialized");
        emit("initialized", null);
    }

    /**
     * Link the current game account with a Discord account via OAuth.
     *
     * @return a future resolving to the linked Discord user ID
     */
    public CompletableFuture<String> linkAccount() {
        requireInitialized();
        log("Account link flow started");
        return CompletableFuture.completedFuture("");
    }

    /**
     * Create or retrieve a provisional (guest) Discord account for the
     * current player, enabling social features without a full Discord login.
     *
     * @return a future resolving to the provisional Discord user ID
     */
    public CompletableFuture<String> getProvisionalAccount() {
        requireInitialized();
        log("Provisional account requested");
        return CompletableFuture.completedFuture("");
    }

    // ──────────────────────────────────────────────
    //  Events
    // ──────────────────────────────────────────────

    /**
     * Subscribe to a named Discord Social event.
     *
     * @param event    event name (e.g. "initialized", "lobbyJoined", "inviteReceived")
     * @param listener callback that receives the event payload
     * @param <T>      the expected payload type
     */
    @SuppressWarnings("unchecked")
    public <T> void on(String event, Consumer<T> listener) {
        listeners.computeIfAbsent(event, k -> new CopyOnWriteArrayList<>())
                .add((Consumer<Object>) listener);
    }

    @SuppressWarnings("unchecked")
    private <T> void emit(String event, T data) {
        List<Consumer<Object>> handlers = listeners.get(event);
        if (handlers != null) {
            for (Consumer<Object> handler : handlers) {
                handler.accept(data);
            }
        }
    }

    // ──────────────────────────────────────────────
    //  Rich Presence
    // ──────────────────────────────────────────────

    /**
     * Set the player's Discord Rich Presence activity text.
     *
     * @param details primary activity description
     * @param state   optional secondary line (e.g. "Round 3 of 10")
     * @return a future that completes when the presence is updated
     */
    public CompletableFuture<Void> setActivity(String details, String state) {
        requireInitialized();
        log("Presence updated — details=\"" + details + "\" state=\"" + (state != null ? state : "") + "\"");
        emit("presenceUpdated", null);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Set Rich Presence party info for multiplayer sessions.
     *
     * @param partyId     unique party identifier
     * @param currentSize number of players currently in the party
     * @param maxSize     maximum party capacity
     * @param joinSecret  optional secret that allows others to join
     * @return a future that completes when the party info is updated
     */
    public CompletableFuture<Void> setParty(String partyId, int currentSize, int maxSize, String joinSecret) {
        requireInitialized();
        log("Party set — id=" + partyId + " size=" + currentSize + "/" + maxSize);
        emit("presenceUpdated", null);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Start an elapsed-time timer on the Rich Presence display.
     *
     * @return a future that completes when the timer is started
     */
    public CompletableFuture<Void> startTimer() {
        requireInitialized();
        log("Presence timer started");
        emit("presenceUpdated", null);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Clear all Rich Presence data.
     *
     * @return a future that completes when the presence is cleared
     */
    public CompletableFuture<Void> clearPresence() {
        requireInitialized();
        log("Presence cleared");
        emit("presenceUpdated", null);
        return CompletableFuture.completedFuture(null);
    }

    // ──────────────────────────────────────────────
    //  Friends
    // ──────────────────────────────────────────────

    /**
     * Retrieve a unified friends list that merges Discord friends with in-game friends.
     *
     * @return a future resolving to the merged friends list
     */
    public CompletableFuture<List<IVXUnifiedFriend>> getUnifiedFriends() {
        requireInitialized();
        log("Fetching unified friends list");
        return CompletableFuture.completedFuture(Collections.emptyList());
    }

    // ──────────────────────────────────────────────
    //  Lobby
    // ──────────────────────────────────────────────

    /**
     * Create or join a lobby identified by a shared secret.
     *
     * @param secret   lobby join secret
     * @param metadata optional key-value metadata for the lobby
     * @return a future resolving to the lobby information
     */
    public CompletableFuture<IVXDiscordLobbyInfo> createOrJoinLobby(String secret, Map<String, String> metadata) {
        requireInitialized();
        IVXDiscordLobbyInfo lobby = new IVXDiscordLobbyInfo(
                "", secret, "", 1, 16, metadata != null ? metadata : Collections.emptyMap());
        this.currentLobby = lobby;
        this.chatHistory.clear();
        log("Lobby joined — secret=" + secret);
        emit("lobbyJoined", lobby);
        return CompletableFuture.completedFuture(lobby);
    }

    /**
     * Leave the current lobby.
     *
     * @return a future that completes when the lobby is left
     */
    public CompletableFuture<Void> leaveLobby() {
        requireInitialized();
        this.currentLobby = null;
        this.chatHistory.clear();
        log("Left lobby");
        emit("lobbyLeft", null);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Send a text-chat message to the current lobby.
     *
     * @param message the message content
     * @return a future that completes when the message is sent
     * @throws IllegalStateException if not in a lobby
     */
    public CompletableFuture<Void> sendMessage(String message) {
        requireInitialized();
        if (currentLobby == null) {
            throw new IllegalStateException("Not in a lobby. Call createOrJoinLobby() first.");
        }
        chatHistory.add(new ChatEntry("self", message, System.currentTimeMillis()));
        log("Chat message sent: " + message);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Return the most recent chat messages for the current lobby.
     *
     * @param limit maximum number of messages to return
     * @return an unmodifiable list of chat entries (newest last)
     */
    public List<ChatEntry> getChatHistory(int limit) {
        requireInitialized();
        int size = chatHistory.size();
        int from = Math.max(0, size - limit);
        return Collections.unmodifiableList(new ArrayList<>(chatHistory.subList(from, size)));
    }

    // ──────────────────────────────────────────────
    //  Voice
    // ──────────────────────────────────────────────

    /**
     * Join a voice call in the specified lobby.
     *
     * @param lobbyId the lobby to join voice in
     * @return a future that completes when the voice call is joined
     */
    public CompletableFuture<Void> joinCall(String lobbyId) {
        requireInitialized();
        log("Joined voice call — lobby=" + lobbyId);
        emit("voiceJoined", lobbyId);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Leave the current voice call.
     *
     * @return a future that completes when the voice call is left
     */
    public CompletableFuture<Void> leaveCall() {
        requireInitialized();
        log("Left voice call");
        emit("voiceLeft", null);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Mute or unmute the local player's microphone.
     *
     * @param muted {@code true} to mute, {@code false} to unmute
     * @return a future that completes when the mute state is updated
     */
    public CompletableFuture<Void> setSelfMute(boolean muted) {
        requireInitialized();
        log("Self mute: " + muted);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Deafen or undeafen the local player.
     *
     * @param deaf {@code true} to deafen, {@code false} to undeafen
     * @return a future that completes when the deafen state is updated
     */
    public CompletableFuture<Void> setSelfDeafen(boolean deaf) {
        requireInitialized();
        log("Self deafen: " + deaf);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Set input (microphone) and output (speaker) volume levels (0–100).
     *
     * @param input  microphone volume
     * @param output speaker volume
     * @return a future that completes when the volume is updated
     */
    public CompletableFuture<Void> setVolume(int input, int output) {
        requireInitialized();
        log("Volume set — input=" + input + " output=" + output);
        return CompletableFuture.completedFuture(null);
    }

    // ──────────────────────────────────────────────
    //  Invites
    // ──────────────────────────────────────────────

    /**
     * Send a game invite to another user by their ID.
     *
     * @param userId  the target user
     * @param message optional invite message
     * @return a future that completes when the invite is sent
     */
    public CompletableFuture<Void> sendInvite(String userId, String message) {
        requireInitialized();
        log("Invite sent to " + userId);
        return CompletableFuture.completedFuture(null);
    }

    /**
     * Register a callback for incoming game invites.
     * Shorthand for {@code on("inviteReceived", handler)}.
     *
     * @param handler callback that receives the invite
     */
    public void onInviteReceived(Consumer<IVXGameInvite> handler) {
        on("inviteReceived", handler);
    }

    /**
     * Register a callback for "Ask to Join" requests.
     * Shorthand for {@code on("joinRequested", handler)}.
     *
     * @param handler callback that receives the requesting user ID
     */
    public void onJoinRequested(Consumer<String> handler) {
        on("joinRequested", handler);
    }

    // ──────────────────────────────────────────────
    //  Internal
    // ──────────────────────────────────────────────

    private void requireInitialized() {
        if (!initialized) {
            throw new IllegalStateException("IVXDiscordSocial has not been initialized. Call initialize() first.");
        }
    }

    private void log(String message) {
        if (config != null && config.isEnableDebugLogs()) {
            LOG.info("[IntelliVerseX:Discord] " + message);
        }
    }

    /**
     * A single chat message entry within a lobby.
     */
    public static final class ChatEntry {
        private final String senderId;
        private final String message;
        private final long timestamp;

        public ChatEntry(String senderId, String message, long timestamp) {
            this.senderId = senderId;
            this.message = message;
            this.timestamp = timestamp;
        }

        public String getSenderId() { return senderId; }
        public String getMessage() { return message; }
        public long getTimestamp() { return timestamp; }

        @Override
        public String toString() {
            return "ChatEntry{senderId='" + senderId + "', message='" + message + "'}";
        }
    }
}

File: SDKs/flutter/lib/src/ivx_discord_social.dart
import 'dart:async';

import 'types.dart';

// ---------------------------------------------------------------------------
// Models
// ---------------------------------------------------------------------------

/// Configuration for the Discord Social SDK integration.
class IVXDiscordConfig {
  final String applicationId;
  final String clientId;
  final String? redirectUri;
  final bool enableDebugLogs;

  const IVXDiscordConfig({
    required this.applicationId,
    required this.clientId,
    this.redirectUri,
    this.enableDebugLogs = false,
  });

  @override
  String toString() =>
      'IVXDiscordConfig(applicationId: $applicationId, clientId: $clientId)';
}

/// A friend entry from the unified (Discord + game) friends list.
class IVXUnifiedFriend {
  final String userId;
  final String? discordId;
  final String username;
  final String displayName;
  final String avatarUrl;

  /// `"discord"`, `"game"`, or `"both"`.
  final String source;

  /// `"online"`, `"idle"`, `"dnd"`, or `"offline"`.
  final String status;

  const IVXUnifiedFriend({
    required this.userId,
    this.discordId,
    required this.username,
    this.displayName = '',
    this.avatarUrl = '',
    this.source = 'game',
    this.status = 'offline',
  });

  factory IVXUnifiedFriend.fromJson(Map<String, dynamic> json) =>
      IVXUnifiedFriend(
        userId: json['user_id'] as String? ?? '',
        discordId: json['discord_id'] as String?,
        username: json['username'] as String? ?? '',
        displayName: json['display_name'] as String? ?? '',
        avatarUrl: json['avatar_url'] as String? ?? '',
        source: json['source'] as String? ?? 'game',
        status: json['status'] as String? ?? 'offline',
      );

  @override
  String toString() =>
      'IVXUnifiedFriend(userId: $userId, username: $username, source: $source)';
}

/// An incoming game invite from another player.
class IVXGameInvite {
  final String inviteId;
  final String senderId;
  final String senderName;
  final String? message;
  final String? lobbyId;
  final DateTime timestamp;

  const IVXGameInvite({
    required this.inviteId,
    required this.senderId,
    required this.senderName,
    this.message,
    this.lobbyId,
    required this.timestamp,
  });

  factory IVXGameInvite.fromJson(Map<String, dynamic> json) => IVXGameInvite(
        inviteId: json['invite_id'] as String? ?? '',
        senderId: json['sender_id'] as String? ?? '',
        senderName: json['sender_name'] as String? ?? '',
        message: json['message'] as String?,
        lobbyId: json['lobby_id'] as String?,
        timestamp: DateTime.tryParse(json['timestamp'] as String? ?? '') ??
            DateTime.now(),
      );

  @override
  String toString() =>
      'IVXGameInvite(inviteId: $inviteId, senderId: $senderId)';
}

/// Information about a Discord lobby.
class IVXDiscordLobbyInfo {
  final String lobbyId;
  final String secret;
  final String ownerId;
  final int memberCount;
  final int maxMembers;
  final Map<String, String> metadata;

  const IVXDiscordLobbyInfo({
    required this.lobbyId,
    required this.secret,
    this.ownerId = '',
    this.memberCount = 1,
    this.maxMembers = 16,
    this.metadata = const {},
  });

  factory IVXDiscordLobbyInfo.fromJson(Map<String, dynamic> json) =>
      IVXDiscordLobbyInfo(
        lobbyId: json['lobby_id'] as String? ?? '',
        secret: json['secret'] as String? ?? '',
        ownerId: json['owner_id'] as String? ?? '',
        memberCount: json['member_count'] as int? ?? 1,
        maxMembers: json['max_members'] as int? ?? 16,
        metadata: (json['metadata'] as Map<String, dynamic>?)
                ?.map((k, v) => MapEntry(k, v.toString())) ??
            const {},
      );

  @override
  String toString() =>
      'IVXDiscordLobbyInfo(lobbyId: $lobbyId, members: $memberCount/$maxMembers)';
}

/// A participant in a Discord voice call.
class IVXVoiceParticipant {
  final String userId;
  final String username;
  final bool isMuted;
  final bool isDeafened;
  final bool isSpeaking;
  final int volume;

  const IVXVoiceParticipant({
    required this.userId,
    required this.username,
    this.isMuted = false,
    this.isDeafened = false,
    this.isSpeaking = false,
    this.volume = 100,
  });

  factory IVXVoiceParticipant.fromJson(Map<String, dynamic> json) =>
      IVXVoiceParticipant(
        userId: json['user_id'] as String? ?? '',
        username: json['username'] as String? ?? '',
        isMuted: json['is_muted'] as bool? ?? false,
        isDeafened: json['is_deafened'] as bool? ?? false,
        isSpeaking: json['is_speaking'] as bool? ?? false,
        volume: json['volume'] as int? ?? 100,
      );

  @override
  String toString() =>
      'IVXVoiceParticipant(userId: $userId, username: $username)';
}

/// A single chat message entry within a lobby.
class IVXChatEntry {
  final String senderId;
  final String message;
  final DateTime timestamp;

  const IVXChatEntry({
    required this.senderId,
    required this.message,
    required this.timestamp,
  });

  @override
  String toString() =>
      'IVXChatEntry(senderId: $senderId, message: $message)';
}

// ---------------------------------------------------------------------------
// Events
// ---------------------------------------------------------------------------

/// Events emitted by [IVXDiscordSocial].
enum IVXDiscordEvent {
  initialized,
  presenceUpdated,
  friendsUpdated,
  lobbyJoined,
  lobbyLeft,
  chatMessage,
  voiceJoined,
  voiceLeft,
  voiceParticipantUpdated,
  inviteReceived,
  joinRequested,
  error,
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/// Discord Social SDK integration for IntelliVerseX.
///
/// Wraps Discord Rich Presence, unified friends list, lobby/text-chat,
/// voice channels, and game invites behind a single ergonomic API.
///
/// ```dart
/// final discord = IVXDiscordSocial.instance;
/// await discord.initialize(
///   config: IVXDiscordConfig(applicationId: '...', clientId: '...'),
/// );
/// await discord.setActivity('Playing QuizVerse', state: 'Round 3');
/// ```
class IVXDiscordSocial {
  static IVXDiscordSocial? _instance;

  IVXDiscordConfig? _config;
  bool _initialized = false;
  IVXDiscordLobbyInfo? _currentLobby;
  final List<IVXChatEntry> _chatHistory = [];

  final _controller = StreamController<MapEntry<IVXDiscordEvent, dynamic>>.broadcast();

  IVXDiscordSocial._();

  /// Singleton accessor.
  static IVXDiscordSocial get instance => _instance ??= IVXDiscordSocial._();

  /// Reset the singleton (useful for testing).
  static void resetInstance() => _instance = null;

  /// Whether [initialize] has been called successfully.
  bool get isInitialized => _initialized;

  /// Stream of all Discord Social events.
  ///
  /// Each event is a [MapEntry] of the [IVXDiscordEvent] and its payload.
  Stream<MapEntry<IVXDiscordEvent, dynamic>> get events => _controller.stream;

  /// Filtered stream for a specific [IVXDiscordEvent].
  Stream<T> on<T>(IVXDiscordEvent event) =>
      _controller.stream
          .where((e) => e.key == event)
          .map((e) => e.value as T);

  // ---------------------------------------------------------------------------
  // Lifecycle
  // ---------------------------------------------------------------------------

  /// Initialize the Discord Social SDK with the given [config].
  ///
  /// Must be called before any other method. Supports account linking and
  /// provisional (guest) accounts.
  Future<void> initialize({required IVXDiscordConfig config}) async {
    if (config.applicationId.trim().isEmpty) {
      throw const IVXError(code: -1, message: 'applicationId cannot be empty.');
    }
    if (config.clientId.trim().isEmpty) {
      throw const IVXError(code: -1, message: 'clientId cannot be empty.');
    }
    _config = config;
    _initialized = true;
    _log('Discord Social SDK initialized');
    _emit(IVXDiscordEvent.initialized, null);
  }

  /// Link the current game account with a Discord account via OAuth.
  ///
  /// Returns the linked Discord user ID on success.
  Future<String> linkAccount() async {
    _ensureInitialized();
    _log('Account link flow started');
    return '';
  }

  /// Create or retrieve a provisional (guest) Discord account for the
  /// current player, enabling social features without a full Discord login.
  Future<String> getProvisionalAccount() async {
    _ensureInitialized();
    _log('Provisional account requested');
    return '';
  }

  // ---------------------------------------------------------------------------
  // Rich Presence
  // ---------------------------------------------------------------------------

  /// Set the player's Discord Rich Presence activity text.
  Future<void> setActivity(String details, {String? state}) async {
    _ensureInitialized();
    _log('Presence updated — details="$details" state="${state ?? ''}"');
    _emit(IVXDiscordEvent.presenceUpdated, null);
  }

  /// Set Rich Presence party info for multiplayer sessions.
  Future<void> setParty(
    String partyId,
    int currentSize,
    int maxSize, {
    String? joinSecret,
  }) async {
    _ensureInitialized();
    _log('Party set — id=$partyId size=$currentSize/$maxSize');
    _emit(IVXDiscordEvent.presenceUpdated, null);
  }

  /// Start an elapsed-time timer on the Rich Presence display.
  Future<void> startTimer() async {
    _ensureInitialized();
    _log('Presence timer started');
    _emit(IVXDiscordEvent.presenceUpdated, null);
  }

  /// Clear all Rich Presence data.
  Future<void> clearPresence() async {
    _ensureInitialized();
    _log('Presence cleared');
    _emit(IVXDiscordEvent.presenceUpdated, null);
  }

  // ---------------------------------------------------------------------------
  // Friends
  // ---------------------------------------------------------------------------

  /// Retrieve a unified friends list that merges Discord friends with
  /// in-game friends.
  ///
  /// Each entry indicates whether the friend was sourced from Discord,
  /// the game backend, or both.
  Future<List<IVXUnifiedFriend>> getUnifiedFriends() async {
    _ensureInitialized();
    _log('Fetching unified friends list');
    return [];
  }

  // ---------------------------------------------------------------------------
  // Lobby
  // ---------------------------------------------------------------------------

  /// Create or join a lobby identified by a shared [secret].
  Future<IVXDiscordLobbyInfo> createOrJoinLobby(
    String secret, {
    Map<String, String>? metadata,
  }) async {
    _ensureInitialized();
    final lobby = IVXDiscordLobbyInfo(
      lobbyId: '',
      secret: secret,
      metadata: metadata ?? const {},
    );
    _currentLobby = lobby;
    _chatHistory.clear();
    _log('Lobby joined — secret=$secret');
    _emit(IVXDiscordEvent.lobbyJoined, lobby);
    return lobby;
  }

  /// Leave the current lobby.
  Future<void> leaveLobby() async {
    _ensureInitialized();
    _currentLobby = null;
    _chatHistory.clear();
    _log('Left lobby');
    _emit(IVXDiscordEvent.lobbyLeft, null);
  }

  /// Send a text-chat [message] to the current lobby.
  Future<void> sendMessage(String message) async {
    _ensureInitialized();
    if (_currentLobby == null) {
      throw const IVXError(
        code: -1,
        message: 'Not in a lobby. Call createOrJoinLobby() first.',
      );
    }
    _chatHistory.add(IVXChatEntry(
      senderId: 'self',
      message: message,
      timestamp: DateTime.now(),
    ));
    _log('Chat message sent: $message');
  }

  /// Return the most recent chat messages for the current lobby.
  List<IVXChatEntry> getChatHistory({int limit = 50}) {
    _ensureInitialized();
    final start = (_chatHistory.length - limit).clamp(0, _chatHistory.length);
    return List.unmodifiable(_chatHistory.sublist(start));
  }

  // ---------------------------------------------------------------------------
  // Voice
  // ---------------------------------------------------------------------------

  /// Join a voice call in the specified lobby.
  Future<void> joinCall(String lobbyId) async {
    _ensureInitialized();
    _log('Joined voice call — lobby=$lobbyId');
    _emit(IVXDiscordEvent.voiceJoined, lobbyId);
  }

  /// Leave the current voice call.
  Future<void> leaveCall() async {
    _ensureInitialized();
    _log('Left voice call');
    _emit(IVXDiscordEvent.voiceLeft, null);
  }

  /// Mute or unmute the local player's microphone.
  Future<void> setSelfMute(bool muted) async {
    _ensureInitialized();
    _log('Self mute: $muted');
  }

  /// Deafen or undeafen the local player.
  Future<void> setSelfDeafen(bool deaf) async {
    _ensureInitialized();
    _log('Self deafen: $deaf');
  }

  /// Set input (microphone) and output (speaker) volume levels (0–100).
  Future<void> setVolume(int input, int output) async {
    _ensureInitialized();
    _log('Volume set — input=$input output=$output');
  }

  // ---------------------------------------------------------------------------
  // Invites
  // ---------------------------------------------------------------------------

  /// Send a game invite to another user by their ID.
  Future<void> sendInvite(String userId, {String? message}) async {
    _ensureInitialized();
    _log('Invite sent to $userId');
  }

  /// Stream of incoming game invites.
  ///
  /// Shorthand for `on<IVXGameInvite>(IVXDiscordEvent.inviteReceived)`.
  Stream<IVXGameInvite> get onInviteReceived =>
      on<IVXGameInvite>(IVXDiscordEvent.inviteReceived);

  /// Stream of "Ask to Join" requests from other players.
  ///
  /// Shorthand for `on<String>(IVXDiscordEvent.joinRequested)`.
  Stream<String> get onJoinRequested =>
      on<String>(IVXDiscordEvent.joinRequested);

  // ---------------------------------------------------------------------------
  // Internal
  // ---------------------------------------------------------------------------

  void _emit(IVXDiscordEvent event, dynamic data) {
    _controller.add(MapEntry(event, data));
  }

  void _ensureInitialized() {
    if (!_initialized) {
      throw const IVXError(
        code: -1,
        message: 'IVXDiscordSocial not initialized. Call initialize() first.',
      );
    }
  }

  void _log(String message) {
    if (_config?.enableDebugLogs ?? false) {
      // ignore: avoid_print
      print('[IntelliVerseX:Discord] $message');
    }
  }
}
